commons/all/src/main/java/org/infinispan/commons/util/concurrent/CompletionStages.java

```diff
@@ -0,0 +1,323 @@
+package org.infinispan.commons.util.concurrent;
+
+
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collector;
+
+/**
+ * Utility methods for handling {@link CompletionStage} instances.
+ * @author wburns
+ * @since 10.0
+ */
+public class CompletionStages {
```

Comment:
@wburns I duplicated this classe because it lives in core
I think we might want to do a refactoring outside this PR

Reply:
Yeah, we should definitely move the core to here instead. Maybe just do it as a separate commit on this PR would be fine.